File: 69_Sqrt_x/mySqrt.py
class Solution:
    def mySqrt(self, x: int) -> int:
        if x == 0: return 0
        left, right = 1, x
        while left < right:
            mid = (left + right) // 2
            if mid ** 2 == x or right - left == 1: return mid
            elif mid ** 2 < x: left = mid
            else: right = mid
        return left

File: 69_Sqrt_x/mySqrt.go
func mySqrt(x int) int {
    if x <= 0 {
        return x
    }

    lo, hi := 1, x
    var mid int
    for lo < hi {
        mid = lo + (hi - lo) >> 1
        if mid * mid == x || hi - lo == 1 {
            return mid
        } else if mid * mid < x {
            lo = mid
        } else {
            hi = mid
        }
    }

    return lo
}
